File: javascript/data-structures/p165_minPQ.js
/*
    CLRS Exercise 6.5-3, p. 165
*/

const { MinHeap } = require('./p156_MinHeap');

class MinPQ extends MinHeap {
    constructor(A, heapSize) {
        super(A, heapSize);
    }

    heapMinimum = () => this.A[0];

    heapExtractMin = () => {
        if (this.heapSize < 1) {
            throw new Error('heap underflow');
        }
        const min = this.heapMinimum();
        this.A[0] = this.A.pop();
        this.heapSize--;
        this.minHeapify(0);
        return min;
    }

    heapDecreaseKey = (i, key) => {
        if (key > this.A[i]) {
            throw new Error('new key is larger than current key');
        }
        this.A[i] = key;
        while (i > 0 && (this.A[this.parent(i)] > this.A[i])) {
            [this.A[i], this.A[this.parent(i)]] = [this.A[this.parent(i)], this.A[i]];
            i = this.parent(i);
        }
    }

    minHeapInsert = (key) => {
        this.heapSize++;
        this.A.push(Number.POSITIVE_INFINITY);
        this.heapDecreaseKey(this.heapSize - 1, key);
    }

    // N.B. cf. Exercise 6.5-8, p. 166
    heapDelete = (i) => {
        // TO-DO
    }
}

module.exports = {
    MinPQ
}

File: python/data_structures/p165_min_pq.py
'''
    CLRS Exercise 6.5-3, p. 165
'''

from .p156_min_heap import MinHeap
import math

class MinPQ(MinHeap):
    def __init__(self, A, heap_size):
        super().__init__(A, heap_size)
    
    def heap_minimum(self):
        return self.A[0]
    
    def heap_extract_min(self):
        if self.heap_size < 1:
            raise Exception('heap underflow')
        min = self.heap_minimum()
        self.A[0] = self.A.pop()
        self.heap_size -= 1
        self.min_heapify(0)
        return min
    
    def heap_decrease_key(self, i, key):
        if key > self.A[i]:
            raise Exception('new key is larger than current key')
        self.A[i] = key
        while i > 0 and self.A[self.parent(i)] > self.A[i]:
            self.A[i], self.A[self.parent(i)] = self.A[self.parent(i)], self.A[i]
            i = self.parent(i)
    
    def min_heap_insert(self, key):
        self.heap_size += 1
        self.A.append(math.inf)
        self.heap_decrease_key(self.heap_size - 1, key)
    
    # N.B. cf. Exercise 6.5-8, p. 166
    def heap_delete(self, i):
        pass # TO-DO
